src/Data/LLVM/BitCode/GetBits.hs

```diff
@@ -125,112 +117,233 @@ instance MonadPlus GetBits where
+--       ws = BS.take rcnt $ BS.drop s8 bs
+
+--       -- Combine the extracted bytes into an Integer value in wi.
+--       wi = BS.foldr (\w a -> a `shiftL` 8 .|. fromIntegral w) (0::Int) ws
+
+--       -- Mask is 0-bit based set of bits wanted in the result
+--       mask = ((1::Int) `shiftL` bitCount numBits) - 1
+
+--       -- Shift the desired value down to byte alignment and then discard any
+--       -- excess high bits.
+--       vi = wi `shiftR` (bitCount startBit .&. 7) .&. mask
+
+--       updPos = addBitCounts startBit numBits
+--   in if updPos > bitLimit
+--      then Left ("Attempt to read bits past limit (newPos="
+--                 <> show updPos <> ", limit=" <> show bitLimit <> ")"
+--                )
+--      else Right (vi, updPos)
+
+extractFromByteString :: Int# {-^ the last bit accessible in the ByteString -}
+                      -> Int# {-^ the bit to start extraction at -}
+                      -> Int# {-^ the number of bits to extract -}
                       -> ByteString {-^ the ByteString to extract from -}
-                      -> Either String (Integer, NumBits)
-extractFromByteString bitLimit startBit numBits bs =
-  let Bytes' s8 = fst (bitsToBytes startBit)
-      Bytes' r8 = fst (bitsToBytes numBits)
-      rcnt = r8 + 2 -- 2 == pre-shift overflow byte on either side
-
-      -- Extract the relevant bits from the ByteCode, with padding to byte
-      -- boundaries into ws.
-      ws = BS.take rcnt $ BS.drop s8 bs
-
-      -- Combine the extracted bytes into an Integer value in wi.
-      wi = BS.foldr (\w a -> a `shiftL` 8 .|. fromIntegral w) (0::Integer) ws
-
-      -- Mask is 0-bit based set of bits wanted in the result
-      mask = ((1::Integer) `shiftL` bitCount numBits) - 1
-
-      -- Shift the desired value down to byte alignment and then discard any
-      -- excess high bits.
-      vi = wi `shiftR` (bitCount startBit .&. 7) .&. mask
+                      -> Either String (() -> (# Int#, Int# #))
+extractFromByteString !bitLim# !sBit# !nbits# bs =
+     if isTrue# ((1# `uncheckedIShiftL#` (nbits#)) /=# 0#)
+        -- (nbits# -# 1#) above would allow 64-bit value extraction, but this
+        -- function cannot actually support a size of 64, because Int# is signed,
+        -- so it doesn't properly use the high bit in numeric operations.  This
+        -- seems to be OK at this point because LLVM bitcode does not attempt to
+        -- encode actual 64-bit values.
+     then
+       let !updPos# = sBit# +# nbits#
+       in if isTrue# (updPos# <=# bitLim#)
+          then
+            let !s8# = sBit# `uncheckedIShiftRL#` 3#
+                !hop# = sBit# `andI#` 7#
+                !r8# = ((hop# +# nbits# +# 7#) `uncheckedIShiftRL#` 3#)
+                !mask# = (1# `uncheckedIShiftL#` nbits#) -# 1#
+
+#if MIN_VERSION_base(4,16,0)
+                word8ToInt !w8# = word2Int# (word8ToWord# w8#)
+#else
+-- technically #if !MIN_VERSION_ghc_prim(0,8,0), for GHC 9.2, but that isn't a define
```

Comment:
More precisely, `MIN_VERSION_ghc_prim` isn't defined because `llvm-pretty-bc-parser` doesn't declare an explicit dependency on `ghc-prim`. But `base`'s `GHC.Exts` does re-export everything of interest from `ghc-prim`, so we could be a little more precise by changing this CPP to `MIN_VERSION_base(4,16,0)`.

Reply:
Ahh, I guess I hadn't tracked the missing dependency/re-export issue.  That said, I'm not sure I understand your suggestion... I think that what I have already?

Reply:
Oops, so it is. In that case, I suppose the only thing left to do is to refine the comment a bit.